### PR TITLE
Tcp collector

### PR DESCRIPTION
**Pull request by e3rd**
Since #1220 cannot be reopened due to Github restrictions, I open a new PR.

Fortunately, I've found somewhere a .pyc file of test_collector that could be decompiled. So the biggest part of the hard work was redeemed.

I've added a test and hereby I submit the work to be reviewed.

### CODE REVIEW

**Comment by e3rd** [comment]
It is very surprising to me that we have different output. If you are still having troubles with passing the tests, tell me your system. Mine's Ubuntu 17.10 Python 3.6.3.

The tests are running smoothly at my system:

```bash
[tcp-collector *$ u=]$ python3 test_collector.py
......
----------------------------------------------------------------------
Ran 6 tests in 6.500s

OK
```

> Because the dict in raw can be sorted differently. The order of elements in dicts is not fixed. So sort the keys in serialization.

I can't sort the dict, nor use assertMessageEqual because the serialization occurs in the bot, not in the tests. However, I changed the line you marked and I'm handling now the raw differently, in the manner that should be indifferent to the dict keys sorting. Could you please provide me full tests output so that I can debug as I am not able to spot the errors?

**Comment by wagner-certat** [comment]
> I can't sort the dict, nor use assertMessageEqual because the serialization occurs in the bot, not in the tests. However, I changed the line you marked and I'm handling now the raw differently, in the manner that should be indifferent to the dict keys sorting.

Seems to work!

With the current code I get the same error on python 3.4 as travis: https://travis-ci.org/certtools/intelmq/jobs/376050516#L658
In the middle of the big block (hard to spot) are these sections:

```
2018-05-11 14:48:13,930 - test-bot - DEBUG - Disconnected from source pipeline.
2018-05-11 14:48:13,930 - test-bot - DEBUG - Disconnected from destination pipeline.
2018-05-11 14:48:13,930 - test-bot - INFO - Bot stopped.
2018-05-11 14:48:13,930 - test-bot - DEBUG - Waiting for incoming message.
2018-05-11 14:48:13,931 - test-bot - DEBUG - Received message {\'feed.accuracy\': 100.0, \'feed.name\': \'Example feed\', \'feed.url\': \'http://localhost/two_files.tar.gz\', \'classification.taxonomy\': \'malicious code\', \'raw\': \'YmFyIHRleHQK\', \'classification.type\': \'c&c\', \'time.observation\': \'2018-03-20T14:05:03+00:00\'}.
2018-05-11 14:48:13,931 - test-bot - ERROR - Reconnecting.
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/bots/outputs/tcp/output.py", line 26, in process
    self.con.sendall(msg)
ConnectionResetError: [Errno 104] Connection reset by peer
2018-05-11 14:48:13,932 - test-bot - INFO - Connected successfully to localhost:5011.
2018-05-11 14:48:13,932 - test-bot - DEBUG - Disconnected from source pipeline.
2018-05-11 14:48:13,932 - test-bot - DEBUG - Disconnected from destination pipeline.
2018-05-11 14:48:13,932 - test-bot - INFO - Bot stopped.
2018-05-11 14:48:13,932 - test-bot - DEBUG - Waiting for incoming message.
2018-05-11 14:48:13,933 - test-bot - DEBUG - Received message {\'feed.accuracy\': 100.0, \'feed.name\': \'Example feed\', \'feed.url\': \'http://localhost/two_files.tar.gz\', \'classification.taxonomy\': \'malicious code\', \'raw\': \'YmFyIHRleHQK\', \'classification.type\': \'c&c\', \'time.observation\': \'2018-03-20T14:05:03+00:00\'}.
2018-05-11 14:48:13,934 - test-bot - ERROR - Bot has found a problem.
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 163, in start
    self.process()
  File "/home/sebastian/dev/intelmq/intelmq/bots/outputs/tcp/output.py", line 35, in process
    self.acknowledge_message()
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 410, in acknowledge_message
    self.__source_pipeline.acknowledge()
  File "/home/sebastian/dev/intelmq/intelmq/lib/pipeline.py", line 243, in acknowledge
    return self.state.get(self.internal_queue, [None]).pop(0)
IndexError: pop from empty list
2018-05-11 14:48:13,934 - test-bot - INFO - Current Message(event): {\'feed.accuracy\': 100.0, \'feed.name\': \'Example feed\', \'feed.url\': \'http://localhost/two_files.tar.gz\', \'classification.taxonomy\': \'malicious code\', \'raw\': \'YmFyIHRleHQK\', \'classification.type\': \'c&c\', \'time.observation\': \'2018-03-20T14:05:03+00:00\'}.
2018-05-11 14:48:13,935 - test-bot - DEBUG - Disconnected from source pipeline.
2018-05-11 14:48:13,935 - test-bot - DEBUG - Disconnected from destination pipeline.
2018-05-11 14:48:13,935 - test-bot - INFO - Bot stopped.
2018-05-11 14:48:13,935 - test-bot - DEBUG - Waiting for incoming message.
2018-05-11 14:48:13,936 - test-bot - DEBUG - Received message {\'feed.accuracy\': 100.0, \'feed.name\': \'Example feed\', \'feed.url\': \'http://localhost/two_files.tar.gz\', \'classification.taxonomy\': \'malicious code\', \'raw\': \'YmFyIHRleHQK\', \'classification.type\': \'c&c\', \'time.observation\': \'2018-03-20T14:05:03+00:00\'}.
```
and then the test hangs.

Other python versions work fine.

If only this test fails on 3.4 but the functionality works fine, we can also skip this test on 3.4.

File: intelmq/bots/collectors/tcp/collector.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                # Read the message data
+                msg = self.recvall(conn, struct.unpack('>I', raw_msglen)[0])
+                report = self.new_report()
+                report.add("raw", msg)
+                self.send_message(report)
+        except socket.error:
+            self.logger.exception("Reconnecting.")
+            self.con.close()
+            self.connect()
+        except AttributeError:
```
Comment (wagner-certat) [inline intelmq/bots/collectors/tcp/collector.py]: What is that for? Only hides error message in my opinion.

Comment (e3rd) [reply intelmq/bots/collectors/tcp/collector.py]: Agree, I did some test cases and in this state of code, I find this superfluous. I added a warning "Incomplete message received" instead.

**Comment by e3rd** [comment]
Thanks for the report! Since the time you wrote the message, I've installed python3.4 and be able to reproduce the bug.  
From ten tests of the method `test_multiple_bots`, half usually ended up with an error, either a hang or an assertion error. I did even a parameter juggling:

```
client_count = 5
msg_count = 100 ⇒ assertion error 413 != 502
msg_count = 200 ⇒ assertion error 913 != 1002
msg_count = 300 ⇒ assertion error 1413 != 1502
msg_count = 400 ⇒ assertion error 1913 OR sometimes 1824 != 2002
```

I really don't know WHY but increasing the new process delay from one to two seconds seemed to make the test reliable. In works now for me for both 3.4 and 3.6, let's wait what Travis'll say.

**Comment by e3rd** [comment]
@wagner-certat , I honestly don't know what to do. After several hours I managed to install Python3.4.8 in a fresh docker and everything still worked alright:
```bash
root@f75ecc0dadaa:/usr/local/lib/python3.4/dist-packages/intelmq/tests/bots/collectors/tcp# python3.4 test_collector.py 
......

--------------------

Ran 6 tests in 9.703s

OK
```

However, Travis has Python3.4.6 which I'm not aware of where to install it from and that still fails. So I rebased the commits and made Python3.4.6 skipped in the tests. 
Travis-CI works now. (Checks failed but because of a `ripencc_abuse_contact` but.)

Do you think you can merge this PR then?

**Comment by wagner-certat** [comment]
I planned to look at it this week to get the RC ready. Thanks for your efforts.

**Pull request merged**